[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked

pianpwk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #175287
* #174629
* __->__ #168052

This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first checking for optimization hints, then using the symbol upper-bound (as a proxy for worst-case redistribution costs), then falling back to a hardcoded value (8192).
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.

Fixes 32 DDE-ing DTensor ops

With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

pianpwk committed:
redistribute cost hints, uneven sharding

[ghstack-poisoned]

pianpwk committed:
Update on "redistribute cost hints, uneven sharding"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "redistribute cost hints, uneven sharding"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "redistribute cost hints, uneven sharding"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first using the symbol upper-bound, then falling back to a hardcoded value (8192), as a proxy for worst-case redistribution costs.
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms, assumes uneven sharding for unbacked sizes, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value doesn't always need them).
- Lastly, testing the effect of size upper bounds on strategy selection required aot_eager (to hardcode redistribute decisions), and this seemed to require enabling proxy tracking during shard prop again. Otherwise, errors like `RuntimeError: Max(1, u2) (<class 'torch.SymInt'>, 140294330350224)is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f98d1b14af0>` shows up for DTensor outer strides...


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first using the symbol upper-bound, then falling back to a hardcoded value (8192), as a proxy for worst-case redistribution costs. The PR currently hardcodes the fallback - wasn't sure if there's a good config to put it in?
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.
- Lastly, testing if specifying upper bounds for unbacked sizes had an effect on matmul strategy selection required using aot_eager backend, so that redistribution decisions were hardcoded. This seemed to require re-enabling proxy tracking during shard prop, basically reverting https://github.com/pytorch/pytorch/pull/163126. Otherwise, errors like `RuntimeError: Max(1, u2) (<class 'torch.SymInt'>, 140294330350224)is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f98d1b14af0>` show up for DTensor outer strides...


With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first using the symbol upper-bound, then falling back to a hardcoded value (8192), as a proxy for worst-case redistribution costs. The PR currently hardcodes the fallback - wasn't sure if there's a good config to put it in?
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.


With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first using the symbol upper-bound, then falling back to a hardcoded value (8192), as a proxy for worst-case redistribution costs. The PR currently hardcodes the fallback - wasn't sure if there's a good config to put it in?
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.


With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first using the symbol upper-bound, then falling back to a hardcoded value (8192), as a proxy for worst-case redistribution costs. The PR currently hardcodes the fallback - wasn't sure if there's a good config to put it in?
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.


With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first using the symbol upper-bound, then falling back to a hardcoded value (8192), as a proxy for worst-case redistribution costs. The PR currently hardcodes the fallback - wasn't sure if there's a good config to put it in?
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.


With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first using the symbol upper-bound, then falling back to a hardcoded value (8192), as a proxy for worst-case redistribution costs. The PR currently hardcodes the fallback - wasn't sure if there's a good config to put it in?
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.


With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first using the symbol upper-bound, then falling back to a hardcoded value (8192), as a proxy for worst-case redistribution costs. The PR currently hardcodes the fallback - wasn't sure if there's a good config to put it in?
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.


With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first using the symbol upper-bound, then falling back to a hardcoded value (8192), as a proxy for worst-case redistribution costs. The PR currently hardcodes the fallback - wasn't sure if there's a good config to put it in?
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.


With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -581,6 +582,88 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
+        from torch.utils._sympy.numbers import int_oo
+        from torch.utils._sympy.reference import PythonReferenceAnalysis
+
+        int_fallback = 8192
```
wasn't sure if there was a config to put this in

pianpwk commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -581,6 +582,88 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
+                PythonReferenceAnalysis,
+                replacements,
+                cost.node.expr.xreplace(replacements),  # type: ignore[arg-type]
+            )
```
subs/xreplace wouldn't simplify to a concrete value, so this seemed like the way?

ezyang commented on torch/distributed/_local_tensor/__init__.py:
```diff
@@ -543,6 +553,11 @@ def ge(self, other: "int | LocalIntNode | ConstantIntNode") -> bool | SymBool:
+    def le(self, other: "int | LocalIntNode | ConstantIntNode") -> bool | SymBool:
+        r = {self._local_ints[r] <= _int_on_rank(other, r) for r in self._local_ints}
+        assert len(r) == 1, (self, other)
+        return torch._C._get_constant_bool_symnode(next(iter(r)))
```
Ah OK, these LocalTensor changes look fine

pianpwk committed:
Update

[ghstack-poisoned]

pianpwk committed:
Update

[ghstack-poisoned]

pianpwk committed:
Update

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/placement_types.py:
```diff
@@ -129,6 +124,35 @@ def _select_split_tensor(
+            for i in range(num_chunks):
+                start = torch.sym_min(split_size * i, dim_size)
+                end = torch.sym_min(split_size * (i + 1), dim_size)
+                chunks.append(tensor.narrow(dim, start, end - start))
```
yea. but is it ok that you don't have that in the tracing branch? i would have thought no

pianpwk replied:
I think this loop slicing should already produce the empty tensors...

pianpwk replied:
yeahh, in the previous impl, chunk would produce k < num_chunks entries, then call `fill_empty_tensor_to_shards` to fill the num_chunks - k rest.

This impl should just have num_chunks calls to narrow, with the later ones being empty slices

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first using the symbol upper-bound, then falling back to a hardcoded value (8192), as a proxy for worst-case redistribution costs.
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.
- Had to do some patching for LocalTensor tests


With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first using the symbol upper-bound, then falling back to a hardcoded value (8192), as a proxy for worst-case redistribution costs.
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.
- Had to do some patching for LocalTensor tests


With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/placement_types.py:
```diff
@@ -182,8 +206,11 @@ def local_shard_size_and_offset(
         Returns (new local shard size, offset)
 
         """
+        from torch.fx.experimental.symbolic_shapes import guard_or_false
```
how come these imports are not done at the top, is it actually a circular import?

pianpwk replied:
there's a `test_not_import_sympy` that avoids importing anything from symbolic_shapes when we first import torch. We end up having to do this quite a lot actually

wconstab commented on torch/distributed/_functional_collectives.py:
```diff
@@ -99,6 +99,24 @@ def is_torchdynamo_compiling():  # type: ignore[misc]
 ]
 
 
+def _chunk_or_narrow_cat(
```
how come we don't use this helper in the change above in _redistribute.py? it seems almost the same but i'm not sure the difference is intentional

wconstab replied:
the one i was referring to was shard_dim_all_to_all_meta

```
    cat_tensor = torch.cat(stacked_list, dim=gather_dim)
    # pyrefly: ignore [unsupported-operation]
    chunk_size = cat_tensor.size(shard_dim) // group_size
    chunk = torch.narrow(cat_tensor, shard_dim, group_rank * chunk_size, chunk_size)
    return chunk.contiguous()
```
though i realized now it is not the same thing.

pianpwk commented on torch/distributed/tensor/_collective_utils.py:
```diff
@@ -173,7 +173,9 @@ def mesh_broadcast(
+    from torch.fx.experimental.symbolic_shapes import guard_or_false
+
+    if guard_or_false(pad_size == 0):
         return tensor
```
fwiw, changing to `statically_known_true(pad_size == 0)` should default to padding in the backed case too

wconstab replied:
can you explain what statically_known_true(pad_size==0) does?

laithsakka commented on torch/_utils.py:
```diff
@@ -1176,9 +1176,12 @@ def _maybe_view_chunk_cat(
     numel_between = math.prod(shape[1:gather_dim]) if gather_dim > 1 else 1
 
-    if shape[0] == group_size and numel_between == 1:
+    if guard_or_false(shape[0] == group_size) and numel_between == 1:
```
guard_or_false around numel_between == 1:?

laithsakka commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -135,6 +137,98 @@ def cache_clear(self):
+    from torch.utils._sympy.numbers import int_oo
+    from torch.utils._sympy.reference import PythonReferenceAnalysis
+
+    int_fallback = 8192
```
This is fine(ish) can we add a BIG TODO that this shall be rewritten to use the optimization_hint API that we will add in symbolic_shapes and link a github issue you can assign it to me i want to track that this is not like this forever

laithsakka commented:
looks fine, lets just create issue to track unification of hinting code.

pianpwk committed:
Update

[ghstack-poisoned]

pianpwk committed:
Update

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first checking for optimization hints, then using the symbol upper-bound (as a proxy for worst-case redistribution costs), then falling back to a hardcoded value (8192).
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.

Fixes 29 DDE-ing DTensor ops

With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update

[ghstack-poisoned]

pianpwk committed:
Update

[ghstack-poisoned]

pianpwk committed:
Update

[ghstack-poisoned]

pianpwk committed:
Update

[ghstack-poisoned]

pianpwk committed:
Update

[ghstack-poisoned]

pianpwk committed:
Update

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first checking for optimization hints, then using the symbol upper-bound (as a proxy for worst-case redistribution costs), then falling back to a hardcoded value (8192).
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.

Fixes 29 DDE-ing DTensor ops

With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk committed:
Update on "[DTensor] assume uneven sharding by default, redistribute cost hints for unbacked"


This extends compiling unbacked matmuls to strategies that require redistribution.

- When negative/zero cost strategies aren't available, cost expressions are functions of shapes, and so are unbacked, and there's no obvious minimum cost choice. The PR follows the inductor size-hinting strategy of using a fallback value, first checking for optimization hints, then using the symbol upper-bound (as a proxy for worst-case redistribution costs), then falling back to a hardcoded value (8192).
- Avoids using `torch.chunk()` in collective implementations, opting for more unbacked-friendly implementations. Chunk is DDE-prone because it calls `split(tensor, chunk_size)` under the hood, which loses track of the original # of chunks, and tries to extract a concrete value from an ugly unbacked expression. In these cases we typically know the # of chunks beforehand (group size), so this seems avoidable.
- Without divisibility axioms for unbacked sizes, assumes that we should handle uneven sharding by default, so pad/narrow ops are run before/after collectives to be divisible by group size (even if the runtime value runs 0 padding or no-op slices). Users can add `torch._check(size % num_shards)` to opt into the faster path.

Fixes 29 DDE-ing DTensor ops

With this PR, going through all 256 input placements (on R/S0/S1/P) of a matmul on a 2d mesh traces through (eager/aot_eager), following this sample script: https://gist.github.com/pianpwk/65a7e3a60354c8dcdbd532b597d96826


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

pianpwk commented:
@pytorchbot merge -i
